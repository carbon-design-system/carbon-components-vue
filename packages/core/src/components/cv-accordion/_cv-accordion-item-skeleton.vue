<template>
  <li :class="[`${carbonPrefix}--accordion__item`, { [`${carbonPrefix}--accordion__item--active`]: open }]">
    <button disabled type="button" :class="`${carbonPrefix}--accordion__heading`">
      <ChevronRight16 :class="`${carbonPrefix}--accordion__arrow`" />
      <cv-skeleton-text :class="`${carbonPrefix}--accordion__title`"></cv-skeleton-text>
    </button>
    <div :class="`${carbonPrefix}--accordion__content`">
      <slot></slot>
    </div>
  </li>
</template>

<script>
import CvSkeletonText from '../cv-skeleton-text/cv-skeleton-text';
import ChevronRight16 from '@carbon/icons-vue/es/chevron--right/16';
import { carbonPrefixMixin } from '../../mixins';

export default {
  name: 'CvAccordionItemSkeleton',
  mixins: [carbonPrefixMixin],
  components: {
    ChevronRight16,
    CvSkeletonText,
  },
  props: {
    open: { type: Boolean, default: false },
  },
};
</script>
