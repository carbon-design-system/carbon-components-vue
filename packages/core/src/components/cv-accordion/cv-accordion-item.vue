<!--
  An item in a Carbon accordion.

  Usage:
    See accordion.vue
-->

<template>
  <li
    data-accordion-item
    class="cv-accordion-item"
    :class="[
      `${carbonPrefix}--accordion__item`,
      {
        [`${carbonPrefix}--accordion__item--active`]: dataOpen,
        [`${carbonPrefix}--accordion__item--${this.animation}`]: animation,
      },
    ]"
    @animationend="onAnimationEnd"
  >
    <button
      ref="button"
      type="button"
      :class="`${carbonPrefix}--accordion__heading`"
      :aria-expanded="dataOpen ? 'true' : 'false'"
      :aria-controls="uid"
      @click="toggle"
    >
      <ChevronRight16 :class="`${carbonPrefix}--accordion__arrow`" />
      <p :class="`${carbonPrefix}--accordion__title`">
        <slot name="title"></slot>
      </p>
    </button>
    <div :id="uid" :class="`${carbonPrefix}--accordion__content`">
      <slot name="content"></slot>
    </div>
  </li>
</template>

<script>
import ChevronRight16 from '@carbon/icons-vue/es/chevron--right/16';
import { uidMixin, carbonPrefixMixin, methodsMixin } from '../../mixins';
export default {
  name: 'CvAccordionItem',
  mixins: [uidMixin, carbonPrefixMixin, methodsMixin({ button: ['blur', 'focus'] })],
  components: { ChevronRight16 },
  props: {
    open: { type: Boolean, default: false },
  },
  watch: {
    open: {
      immediate: true,
      handler(value) {
        this.dataOpen = value;
      },
    },
  },
  data() {
    return {
      animation: '',
      dataOpen: false,
    };
  },
  mounted() {
    this.$_CvAccordionItem = true; // for use by parent with $children
  },
  methods: {
    toggle(force) {
      this.animation = this.dataOpen ? 'collapsing' : 'expanding';
      const newValue = typeof force === 'boolean' ? !!force : !this.dataOpen;
      const change = this.dataOpen !== undefined && newValue !== this.dataOpen;

      this.dataOpen = newValue;
      if (change) {
        this.$parent.$emit('cv:change', this);
      }
    },
    onAnimationEnd() {
      this.animation = '';
    },
  },
};
</script>
