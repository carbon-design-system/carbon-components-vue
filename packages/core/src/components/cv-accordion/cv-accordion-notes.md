# CvAccordion

A Vue implementation of a Carbon Components accordion.

http://www.carbondesignsystem.com/components/accordion/code

## Usage

CvAccordion is used in conjunction with slotted CvAccordionItem components.

```html
<cv-accordion>
  <cv-accordion-item>
    <template slot="title">Section 1 title </template>
    <template slot="content">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat.
      </p>
    </template>
  </cv-accordion-item>
  <cv-accordion-item>
    <template slot="title">Section 2 title</template>
    <template slot="content">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat.
      </p>
    </template>
  </cv-accordion-item>
  <cv-accordion-item>
    <template slot="title">Section 3 title</template>
    <template slot="content">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat.
      </p>
    </template>
  </cv-accordion-item>
  <cv-accordion-item>
    <template slot="title">Section 4 title</template>
    <template slot="content">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat.
      </p>
    </template>
  </cv-accordion-item>
</cv-accordion>
```

## Events

CvAccordion

- change: when an accordion item opens or closes

## Attributes

CvAccordion
align: Accordion heading alignment (start, end)
size: default, 'sm' or 'xl'

CvAccordionItem
disabled: accordion item disabled
open: accordion item is open

### Additional

N/A
