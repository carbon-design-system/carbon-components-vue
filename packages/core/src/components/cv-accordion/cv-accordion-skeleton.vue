<template>
  <ul :class="`${carbonPrefix}--accordion ${carbonPrefix}--skeleton`">
    <cv-accordion-item-skeleton :open="true">
      <cv-skeleton-text width="90%"></cv-skeleton-text>
      <cv-skeleton-text width="80%"></cv-skeleton-text>
      <cv-skeleton-text width="95%"></cv-skeleton-text>
    </cv-accordion-item-skeleton>
    <cv-accordion-item-skeleton></cv-accordion-item-skeleton>
    <cv-accordion-item-skeleton></cv-accordion-item-skeleton>
    <cv-accordion-item-skeleton></cv-accordion-item-skeleton>
  </ul>
</template>

<script>
import CvSkeletonText from '../cv-skeleton-text/cv-skeleton-text';
import CvAccordionItemSkeleton from './_cv-accordion-item-skeleton';
import { carbonPrefixMixin } from '../../mixins';

export default {
  name: 'CvAccordionSkeleton',
  mixins: [carbonPrefixMixin],
  components: {
    CvSkeletonText,
    CvAccordionItemSkeleton,
  },
};
</script>
