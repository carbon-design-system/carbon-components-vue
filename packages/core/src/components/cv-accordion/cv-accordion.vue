<template>
  <ul data-accordion :class="`cv-accordion ${carbonPrefix}--accordion`">
    <slot></slot>
  </ul>
</template>

<script>
import { carbonPrefixMixin } from '../../mixins';

export default {
  name: 'CvAccordion',
  mixins: [carbonPrefixMixin],
  created() {
    this.$on('cv:change', srcComponent => this.onCvChange(srcComponent));
  },
  computed: {
    state: {
      get() {
        const items = this.$children.filter(item => item.$_CvAccordionItem);
        return items.map(item => item.dataOpen);
      },
      set(val) {
        const items = this.$children.filter(item => item.$_CvAccordionItem);

        if (items.length > 0) {
          // loop through all items setting
          for (const i in items) {
            // making no attempt to verify array
            items[i].dataOpen = val[i] !== undefined ? !!val[i] : false;
          }
        }
      },
    },
  },
  methods: {
    onCvChange(srcComponent) {
      const index = this.$children.findIndex(item => item.uid === srcComponent.uid);
      this.$emit('change', { changedIndex: index, state: this.state });
    },
  },
};
</script>
