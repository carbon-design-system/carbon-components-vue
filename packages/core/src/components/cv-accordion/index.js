import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
