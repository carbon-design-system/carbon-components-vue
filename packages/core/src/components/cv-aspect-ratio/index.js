import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
