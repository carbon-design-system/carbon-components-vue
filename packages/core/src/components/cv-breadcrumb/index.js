import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
