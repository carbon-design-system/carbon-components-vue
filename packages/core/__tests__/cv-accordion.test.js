// import { shallowMount as shallow } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, trigger } = awaitNextTick;
import { CvAccordion, CvAccordionItem, CvAccordionSkeleton } from '@/components/cv-accordion';
import { settings as carbonSettings } from 'carbon-components';

const carbonPrefix = carbonSettings.prefix;

describe('CvAccordion', () => {
  // ***************
  // PROP CHECKS
  // ***************

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('matches render', async () => {
    const wrapper = await shallow(CvAccordion, {
      slots: {
        default: '<div class="cv-accordion-item-stub">StubbyMcStubFace</div>',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
});

describe('CvAccordionSkeleton', () => {
  // ***************
  // PROP CHECKS
  // ***************

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('matches render', async () => {
    const wrapper = await shallow(CvAccordionSkeleton);

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
});

describe('CvAccordionItem', () => {
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvAccordionItem, ['open'], Boolean);

  testComponent.propsHaveDefault(CvAccordionItem, ['open']);

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('matches render', async () => {
    const wrapper = await shallow(CvAccordionItem, {
      propsData: { id: '1' },
      slots: {
        default: '<div class="cv-accordion-item-stub">StubbyMcStubFace</div>',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('matches render open', async () => {
    const wrapper = await shallow(CvAccordionItem, {
      propsData: {
        id: '1',
        open: true,
      },
      slots: {
        default: '<div class="cv-accordion-item-stub">StubbyMcStubFace</div>',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('closed to open to closed', async () => {
    const wrapper = await shallow(CvAccordionItem, {
      slots: {
        default: '<div class="cv-accordion-item-stub">StubbyMcStubFace</div>',
      },
    });

    const button = wrapper.find('button');

    expect(button.attributes('aria-expanded')).toEqual('false');
    expect(wrapper.classes()).not.toContain(`${carbonPrefix}--accordion__item--active`);

    await trigger(wrapper.find('button'), 'click');

    expect(button.attributes('aria-expanded')).toEqual('true');
    expect(wrapper.classes()).toContain(`${carbonPrefix}--accordion__item--active`);

    await trigger(wrapper.find('button'), 'click');

    expect(button.attributes('aria-expanded')).toEqual('false');
    expect(wrapper.classes()).not.toContain(`${carbonPrefix}--accordion__item--active`);
  });

  it('topen to closed to open', async () => {
    const wrapper = await shallow(CvAccordionItem, {
      propsData: {
        open: true,
      },
      slots: {
        default: '<div class="cv-accordion-item-stub">StubbyMcStubFace</div>',
      },
    });

    const button = wrapper.find('button');

    expect(button.attributes('aria-expanded')).toEqual('true');
    expect(wrapper.classes()).toContain(`${carbonPrefix}--accordion__item--active`);

    await trigger(wrapper.find('button'), 'click');

    expect(button.attributes('aria-expanded')).toEqual('false');
    expect(wrapper.classes()).not.toContain(`${carbonPrefix}--accordion__item--active`);

    await trigger(wrapper.find('button'), 'click');

    expect(button.attributes('aria-expanded')).toEqual('true');
    expect(wrapper.classes()).toContain(`${carbonPrefix}--accordion__item--active`);
  });
});
