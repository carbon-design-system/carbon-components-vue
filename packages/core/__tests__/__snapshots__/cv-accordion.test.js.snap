// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvAccordion matches render 1`] = `
<ul data-accordion="" class="cv-accordion bx--accordion bx--accordion--start">
  <div class="cv-accordion-item-stub">StubbyMcStubFace</div>
</ul>
`;

exports[`CvAccordionItem matches render 1`] = `
<li data-accordion-item="" class="cv-accordion-item bx--accordion__item"><button type="button" aria-expanded="false" aria-controls="1" class="bx--accordion__heading">
    <chevronright16-stub class="bx--accordion__arrow"></chevronright16-stub>
    <p class="bx--accordion__title"></p>
  </button>
  <div id="1" class="bx--accordion__content"></div>
</li>
`;

exports[`CvAccordionItem matches render open 1`] = `
<li data-accordion-item="" class="cv-accordion-item bx--accordion__item bx--accordion__item--active"><button type="button" aria-expanded="true" aria-controls="1" class="bx--accordion__heading">
    <chevronright16-stub class="bx--accordion__arrow"></chevronright16-stub>
    <p class="bx--accordion__title"></p>
  </button>
  <div id="1" class="bx--accordion__content"></div>
</li>
`;

exports[`CvAccordionSkeleton matches render 1`] = `
<ul class="bx--accordion bx--skeleton bx--accordion--start">
  <cv-accordion-item-skeleton-stub open="true">
    <cv-skeleton-text-stub width="90%" linecount="3"></cv-skeleton-text-stub>
    <cv-skeleton-text-stub width="80%" linecount="3"></cv-skeleton-text-stub>
    <cv-skeleton-text-stub width="95%" linecount="3"></cv-skeleton-text-stub>
  </cv-accordion-item-skeleton-stub>
  <cv-accordion-item-skeleton-stub></cv-accordion-item-skeleton-stub>
  <cv-accordion-item-skeleton-stub></cv-accordion-item-skeleton-stub>
  <cv-accordion-item-skeleton-stub></cv-accordion-item-skeleton-stub>
</ul>
`;
